### PR TITLE
[OSC-966] fix bugs, add unittesting, minor linting

### PR DESCRIPTION
**Pull request by seanbudd**
https://pageuppeople.atlassian.net/browse/OSC-966

**Main bug:**
_Former logic_: If we haven't lost track of changes in the mssql database (eg due to retention time limits), we updated a flag from 0 to the tracking number. Later, if the tracking number had remained the initial set value of 0, we would perform a full refresh.

_Issue_: If nothing has changed in a table since change tracking has been enabled, then the tracking number is 0. 

_Solution_: Update a boolean flag using the former condition, that way it is not dependent on the tracking number being >0

**Unit testing:**
update `modules\tests\config\connection.json` as appropriate with your mssql database credentials (for master)
run `python run_tests.py`  to perform the unit tests

### CODE REVIEW

**Comment by seanbudd** [comment]
@pep8speaks resume

File: modules/data_sources/MsSqlDataSource.py [hunk 1]
```diff
@@ -108,36 +103,63 @@ def get_next_data_frame(self, table_configuration, columns, batch_configuration,
     def init_change_tracking(self, table_configuration, last_sync_version):
 
         sql_builder = io.StringIO()
-        sql_builder.write(
-            "IF NOT EXISTS(SELECT 1 FROM sys.change_tracking_tables WHERE object_id = OBJECT_ID('{0}.{1}'))\n".format(
-                table_configuration['schema'], table_configuration['name']))
+        sql_builder.write("IF NOT EXISTS(SELECT 1 FROM sys.change_tracking_tables WHERE "
+                          f"object_id = OBJECT_ID('{table_configuration['schema']}.{table_configuration['name']}'))\n")
```
Comment (elexisvenator) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
                          f"object_id = OBJECT_ID('{table_configuration['schema']}.{table_configuration['name']}'));\n")
```

File: modules/data_sources/MsSqlDataSource.py [hunk 2]
```diff
@@ -41,31 +41,25 @@ def build_select_statement(self, table_configuration, columns, batch_configurati
...
+            sql_builder.write("chg.SYS_CHANGE_VERSION as data_pipeline_change_version, "
+                              "CASE chg.SYS_CHANGE_OPERATION WHEN 'D' THEN 1 ELSE 0 END as data_pipeline_is_deleted \n")
+            sql_builder.write(f" FROM CHANGETABLE(CHANGES"
+                              f" {table_configuration['schema']}.{table_configuration['name']},"
+                              f" {change_tracking_info.this_sync_version})"
+                              f" AS chg")
+            sql_builder.write(f" LEFT JOIN {table_configuration['schema']}.{table_configuration['name']} t"
+                              f" on {self.build_change_table_on_clause(batch_key_tracker)}")
+            sql_builder.write(f" WHERE {self.build_where_clause(batch_key_tracker, 'chg')}")
+            sql_builder.write(f" ORDER BY {order_by}")
```
Comment (elexisvenator) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
            sql_builder.write(f" ORDER BY {order_by};")
```

File: modules/data_sources/MsSqlDataSource.py [hunk 3]
```diff
@@ -41,31 +41,25 @@ def build_select_statement(self, table_configuration, columns, batch_configurati
...
-                                                                                       table_configuration[
-                                                                                           'schema'],
-                                                                                       table_configuration[
-                                                                                           'name'],
-                                                                                       self.build_where_clause(batch_key_tracker, "t"),
-                                                                                       order_by)
+            return f"SELECT TOP ({batch_configuration['size']}) {column_names} " \
+                f"FROM {table_configuration['schema']}.{table_configuration['name']} t " \
+                f"WHERE {self.build_where_clause(batch_key_tracker, 't')} " \
+                f"ORDER BY {order_by}"
```
Comment (elexisvenator) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
                f"ORDER BY {order_by};"
```

File: integration_tests/mssql_source/source_database_setup/create_database.sql [hunk 4]
```diff
@@ -2,4 +2,4 @@ IF NOT EXISTS (SELECT * FROM sys.databases WHERE Name = 'RelationalDataLoaderInt
 	CREATE DATABASE	 RelationalDataLoaderIntegrationTestSource
 
 ALTER DATABASE RelationalDataLoaderIntegrationTestSource
-SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 2 DAYS, AUTO_CLEANUP = ON)
\ No newline at end of file
+SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 2 DAYS, AUTO_CLEANUP = ON)
```
Comment (elexisvenator) [inline integration_tests/mssql_source/source_database_setup/create_database.sql]:
```suggestion
SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 2 DAYS, AUTO_CLEANUP = ON);
```

File: modules/tests/setup/create_simple_table.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+INSERT [SimpleTest]
+(
+        Id,
+        StringCol
+)
+SELECT	1,'Frank'
+UNION ALL
+SELECT	2,'Walker'
+UNION ALL
+SELECT	3,'National'
```
Comment (elexisvenator) [inline modules/tests/setup/create_simple_table.sql]:
```suggestion
SELECT	3,'National';
```

File: modules/tests/setup/create_simple_table.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
+IF object_id('SimpleTest') IS NOT NULL
+	DROP TABLE SimpleTest
```
Comment (elexisvenator) [inline modules/tests/setup/create_simple_table.sql]:
```suggestion
	DROP TABLE SimpleTest;
```

File: modules/tests/test_MsSqlDataSource.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                set_up_db_string = f.read()
+                temp_eng = create_engine(gen_connection_string.format(db=file_obj["db"]),
+                                         connect_args={'autocommit': True})
+                temp_eng.execute(set_up_db_string.format(db=TEST_DB))
+
+    @classmethod
+    def tearDownClass(cls):
+        TEAR_DOWN_STRING = """
+            USE [master];
+            IF EXISTS (SELECT * FROM sys.databases WHERE Name = '{db}')
```
Comment (elexisvenator) [inline modules/tests/test_MsSqlDataSource.py]:
```suggestion
            IF EXISTS (SELECT * FROM sys.databases WHERE Name = '{db}';)
```

**Comment by elexisvenator** [comment]
Make sure you fix what is breaking in AppVeyor

File: run_tests.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from modules.tests import test_MsSqlDataSource
+import unittest
+
+suite = unittest.TestLoader().loadTestsFromModule(test_MsSqlDataSource)
+unittest.TextTestRunner(verbosity=2).run(suite)
```
Comment (waltaro) [inline run_tests.py]: What level of verbosity does `2` represent?

Comment (seanbudd) [reply run_tests.py]: added the following comment: https://stackoverflow.com/a/1322648/8030743

File: run_tests.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from modules.tests import test_MsSqlDataSource
+import unittest
+
+suite = unittest.TestLoader().loadTestsFromModule(test_MsSqlDataSource)
+# https://stackoverflow.com/a/1322648/8030743
+unittest.TextTestRunner(verbosity=2).run(suite)
```
Comment (waltaro) [inline run_tests.py]: Can we do something like `LEVEL_VERBOSE = 2`?

**Pull request merged**